<html>
<head>
	<title>
		Human genome assembly project documentation
	</title>
</head>
<body>
	<h1>PROJECT DOCUMENTATION</h1>
	<h2>Design</h2>
	<h3>CRC cards</h3>
		<img src='images/CRCcards.png ' width = 950 height = 500/>
	<h3>UML</h3>
		<img src='images/UMLimage.png'width = 800 height = 500/>
	<h2>Implementation</h2>
	<h3>Reader</h3>
		<p>
		Following the project specifications, we implemented the abstract interface <b>DatasetReader</b>, which presents the abstract method <i><b>read()</i></b>, overridden by the <i><b>read()</i></b> method of the <b>Gff3Reader</b>, the DatasetReader subclass. 
		</p>
		<p>
		Gff3Reader is thus the realisation of the DatasetReader abstract interface; it is a concrete class that accepts a tabular file of format gff3 as an input and returns a Dataset object. We made sure that the class Gff3Reader was specific for the gff3 format by requesting that the fiel directory ended with '.gff3'. We analysed the file directory by splitting the whole using ‘.’ as delimiter, considering that the specification of the file type is contained in the last string, <i>e.g.</i> 
		<ul>
			<li>‘Auto.tsv’ ->  last string = file type = ‘tsv’ -> raises exception</li>
			<li>‘Homo_sapiens.GRCh38.85.gff3’ -> last string = file type = ‘gff3’ -> works</li>
		</ul>
		</p>
	<h3>Dataset and operations</h3>
		<p>
		The Dataset class is built around a Pandas Dataframe and can perform a list of active operations on that, which are listed as its public methods; all of those operations return a Dataset object. 
		</p>
		<p>
		The only two private methods found in the Dataset class are: 
		<ul>
			<li><b><i>__uniqueList(column)</b></i> – given a column label, it returns the unique values in it in the form of a Pandas DataFrame;
				implemented to exploit reusability of code, as it is utilised by both <i><b>uniqueID()</i></b> and <i><b>uniqueType()</i></b>.</li>
			<li><b><i>__uniqueCount(column)</b></i> – given a column label, it returns the count of each of its values. Returns a Pandas DataFrame.
				As above, implemented to avoid repetition of code; it is utilised by <i><b>countSource()</b></i> and <i><b>countType()</i></b>.</li>
		</ul>
		</p>
		<p>
		As for the other methods found, which return all a Dataset object:
		<ul>
			<li><b><i>getDf()</b></i> – retrieves the private attribute <i><b>self.__df</i></b> (type Pandas DataFrame);</li>
			<li><b><i>basicInfo()</b></i> – returns a Dataset object which provides us the column names and the type of the data contained in each of them; i.e. it can be used to retrieve the basic information on the Dataset obtained from the gff3 file reading;</li>
			<li><b><i>uniqueID()</b></i> – returns a Dataset object of unique values of column seqID (no duplicates);</li>
			<li><b><i>uniqueType()</b></i> – returns a Dataset object which informs us on the list of unique type of operations available in the dataset;</li>
			<li><b><i>countSource()</b></i> – returns a Dataset object with the count of Source column – for each source, how many entries do we have of it?</li>
			<li><b><i>countType()</b></i> – returns a Dataset object wuth the count of Type column – for each type, how many entries do we have? </li>
			<li><b><i>entireChromosome()</b></i> – filters the column source to identify only entire chromosomes rows, which are returned as a Dataset object – <i>i.e.</i> only rows with source GRCh38 are selected and returned as a new dataset object;</li>
			<li><b><i>unassembledSequence()</b></i> – returns a Dataset object with the fraction and the percentage of the unassembled sequences over the dataset provided by entireChromosome(), retrieving first the Dataset obtained by ‘<b><i>entireChromosome()</b></i>’ method and then selecting only the entries of type ‘superconting’; eventually, it performs the fraction of the unassembled over the total entries to obtain the ratio, which is expressed also as percentage;</li>
			<li><b><i>onlyEnsemblHavana()</b></i> – returns a Dataset object with only entries from sources ensembl, havana and ensembl_havana;</li>
			<li><b><i>entriesEnsemblHavana()</b></i> – returns a Dataset object with the counts of the number of entries for each unique type found in the <i><b>onlyEnsemblHavana()</b></i> Dataset;</li>
			<li><b><i>ensemblHavanaGenes()</b></i> – returns a Dataset object with the name of the genes present in the Dataset obtained by <i><b>onlyEnsemblHavana()</b></i>.</li>
		</ul>
		</p>
	<h3>Decorator</h3>
		<p>
		The decorator implemented in our program reads a .csv file containing the name of a function and if it is active or not (True or False).
		<ul>
			<li>If the operation is not active, when the user tries to execute it in the web interface it gives an Internal Server error </li>
			<li>If the operation is active, the user sees the returning Dataset printed in a table.</li>
		</ul>	
		</p>
		<p>
			A note on <b>active operations</b> and the <b>decorator</b> that manages them: 
			our choice of implementation was to use a csv file to compile the list of active operation, and a function decorator to manage them. This meant that our UML class diagram could not figure nor the registry of active operation, nor the decorator, albeit they both interact with the class Dataset, since it must check for an operation to be active before performing them. For coherence, and as we found it to be rather disorienting, we also did not add the collaboration to the CRC cards.
		</p>
	<h3>Web-based user interface</h3>
		<p>
		The software functions through the execution of the main module implemented with Flask. There we read the .gff3 file using the Gff3Reader. Using the link the user visualizes an hompegage that contains two links: 
		<ul>
			<li>one for the <b>project documentation</b> with CRC cards and UML; </li>
			<li>one for the <b>operations</b> to perform on the Dataset: opens a list the operations, each one with a link that opens in a new tab the table returned by the operation. </li>
		</ul>
		</p>
		<p>
		The web implementation is written in .html files that are then easily integrated with Flask using ‘render_template’.
		</p>
	
</body>
</html>
